### PR TITLE
Catch exception if git is not installed

### PR DESCRIPTION
**Pull request by noiioiu**
This is a trivial pr that should fix a bug in `cabal init` when `git` is not installed.

---

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.
* [x] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added.

## QA Notes

`cabal init` should work correctly even if `git` is not installed. This can be tested by running the following commands in an empty directory and answering the prompts:

```
mkdir /tmp/bin
ln -s "$(which ghc)" "$(which cabal)" /tmp/bin/
PATH=/tmp/bin/ cabal init
```

### CODE REVIEW

**Comment by geekosaur** [comment]
Could you fill out one of the templates, please? I don't think this needs a changelog fragment, though.

**Comment by ffaf1** [comment]
I wonder if there is a way to test this automatically (if not, just user testing will do).

**Comment by geekosaur** [comment]
Yes, I think for this one we want (very simple) manual QA notes.

**Comment by noiioiu** [comment]
> Could you fill out one of the templates, please? I don't think this needs a changelog fragment, though.

Sure, do you have a link to the templates? I can't find them.

**Comment by geekosaur** [comment]
https://github.com/haskell/cabal/blob/master/.github/pull_request_template.md

**Comment by geekosaur** [comment]
Hm, building is failing https://github.com/haskell/cabal/actions/runs/11562678296/job/32184393856?pr=10486#step:6:1363

**Comment by geekosaur** [comment]
That's not going to work because `m` isn't guaranteed to have access to `Exception` (`cabal init -n` doesn't). I _think_ you will have to arrange to run something like `"sh", ["-c", "git … || exit 0"]` and deal with the output being empty.

**Comment by noiioiu** [comment]
Right, I don't think it's possible to define a MonadCatch instance for PurePrompt.

**Comment by ulysses4ever** [comment]
> Yes, I think for this one we want (very simple) manual QA notes. I wouldn't trust trying to hide or remove git from a runner environment.

Why not? It should be as simple as the description of this PR shows: "simply" resetting the PATH. I think the testsuite might have a combinator for manipulating PATH. 

I don't want to block it for this reason, but having a regression test would be very valuable in this case.

**Comment by noiioiu** [comment]
Added tests to the checklist. I'm not sure how or where to add the test though.

**Review by ulysses4ever — CHANGES_REQUESTED** [review]
Thanks for reporting and taking this on!

An integration test manipulating PATH can be found here: https://github.com/haskell/cabal/blob/63c486a1a65de599fa435c0cbf11ad85725f3c6c/cabal-testsuite/PackageTests/BuildTools/Foreign/setup.test.hs

General setup for integration tests is described here: https://github.com/haskell/cabal/blob/63c486a1a65de599fa435c0cbf11ad85725f3c6c/cabal-testsuite/README.md

Good luck!

**Comment by geekosaur** [comment]
My worry with `$PATH` is that, on Linux at least, the part that needs to be removed is `/usr/bin`.

File: cabal-install/src/Distribution/Client/Init/Types.hs [hunk 1]
```diff
@@ -438,6 +441,7 @@ instance Interactive PurePrompt where
   readProcessWithExitCode !_ !_ !_ = do
     input <- pop
     return (ExitSuccess, input, "")
+  maybeReadProcessWithExitCode a b c = Just <$> readProcessWithExitCode a b c
```
[the remark this answers is not in the document]
Comment (noiioiu) [reply cabal-install/src/Distribution/Client/Init/Types.hs:444]: It should, the `readProcessWithExitCode` used in the definition of `maybeReadProcessWithExitCode`  is the same one defined on line 441, right?

[the remark this answers is not in the document]
Comment (ulysses4ever) [reply cabal-install/src/Distribution/Client/Init/Types.hs:444]: oh, my bad!

**Review by ulysses4ever — APPROVED** [review]
Brilliant! 

Have you tried to run the test without your patch? That would be a good sanity check to see that it fails since it's a somewhat non-trivial test.

**Comment by noiioiu** [comment]
> Brilliant!
> 
> Have you tried to run the test without your patch? That would be a good sanity check to see that it fails since it's a somewhat non-trivial test.

I ran it without the patch on a different branch (forked from 3.10) since I couldn't build `master` on my computer. It failed with the `posix_spawnp` error.

**Review by geekosaur — APPROVED** [review]

**Comment by Kleidukos** [comment]
Assigning myself for manual QA once this lands on `master`

**Comment by noiioiu** [comment]
> > since I couldn't build `master` on my computer.
> 
> Uh, what happened? Is this something likely to affect other people?

I think it might be because I was using cabal 3.10.3.0.

> I assume the `posix_spawn` is what we expect to see without the patch (as reported in #10484) --- that's great!
> 

Yes, it's the same error.

**Comment by geekosaur** [comment]
Older `cabal-install` should be able to install newer ones, so that sounds like a bug. Please file a new issue including as many details about what you did and what happened as you can.

**Comment by ulysses4ever** [comment]
@Mergify backport 3.14

**Comment by geekosaur** [comment]
@mergify backport 3.12

**Pull request merged**